src/common/Formatter.cc

```diff
@@ -366,9 +366,24 @@ std::ostream& JSONFormatter::dump_stream(std::string_view name)
 void JSONFormatter::dump_format_va(std::string_view name, const char *ns, bool quoted, const char *fmt, va_list ap)
 {
   char buf[LARGE_SIZE];
-  vsnprintf(buf, LARGE_SIZE, fmt, ap);
+  char *final_buf = buf;
+  int buf_size = LARGE_SIZE;
+
+  va_list ap_copy;
+  va_copy(ap_copy, ap);
+  int len = vsnprintf(final_buf, buf_size, fmt, ap_copy);
+  va_end(ap_copy);
+
+  std::unique_ptr<char[]> large_buf;
+  if (len >= buf_size) {
+    // output was truncated, allocate a buffer large enough
+    buf_size = len + 1;
+    large_buf = std::make_unique<char[]>(buf_size);
+    vsnprintf(large_buf.get(), buf_size, fmt, ap);
+    final_buf = large_buf.get();
```

Comment:
i really like boost's [small_vector](https://www.boost.org/doc/libs/1_85_0/doc/html/boost/container/small_vector.html#id-1.3.9.14.37.8.4) for things like this. it has the same stack vs. heap behavior, but saves you from having to manage them separately. what do you think?
```c++
  auto buf = boost::container::small_vector<char, LARGE_SIZE>{
      LARGE_SIZE, boost::container::default_init};

  va_list ap_copy;
  va_copy(ap_copy, ap);
  int len = vsnprintf(buf.data(), buf.size(), fmt, ap_copy);
  va_end(ap_copy);

  if (std::cmp_greater_equal(len, buf.size())) {
    // output was truncated, allocate a buffer large enough
    buf.resize(len + 1, boost::container::default_init);
    vsnprintf(buf.data(), buf.size(), fmt, ap);
  }
```

Reply:
Thanks for your kindness. I've applied the suggestion and found another HTMLFormatter that needs fixing. The unittests have also been added. @cbodley